src/utils/board/board.ts

```diff
@@ -0,0 +1,207 @@
+    const secondAllowedDirection = boardSize + 1;
+
+    return [firstAllowedDirection, secondAllowedDirection];
+};
+
+const isShouldBecomeKing = (currentPiece: ICell, selectedPiece: ICell): boolean => {
+    if (!currentPiece) {
+        return false;
+    }
+    if (selectedPiece.row === 0 && currentPiece.piece?.color === PieceColor.White) {
+        return true;
+    }
+
+    return selectedPiece.row === BOARD_SIZE - 1 && currentPiece.piece?.color === PieceColor.Black;
+};
+
+const movePiece = (
+    board: ICell[],
+    currentPiece: ICell,
+    selectedPiece: ICell,
+    updateBoard: boolean,
+    activeSide: PieceColor,
+    setBoard: (board: ICell[]) => void,
+    setCurrentPiece: (piece: ICell | null) => void,
+    setUpdateBoard: (updateBoard: boolean) => void,
+    setActiveSide: (activeSide: PieceColor) => void,
+): void => {
+    const newBoard = Object.assign(board);
+    const oldPosition = newBoard.findIndex((item: ICell) => (
+        item.row === currentPiece?.row && item.col === currentPiece?.col)
+    );
+    const newPosition = newBoard.findIndex((item: ICell) => (
+        item.row === selectedPiece?.row && item.col === selectedPiece?.col)
+    );
+
+    newBoard[newPosition] = {
+        ...newBoard[newPosition],
+        piece: newBoard[oldPosition].piece,
+    };
+
+    newBoard[oldPosition] = {
+        ...newBoard[oldPosition],
+        piece: null,
+    };
+
+    if (
+        currentPiece.piece?.state !== PieceState.King &&
+        isShouldBecomeKing(currentPiece, selectedPiece)
+    ) {
+        newBoard[newPosition] = {
+            ...newBoard[newPosition],
+            piece: {
+                ...newBoard[newPosition].piece,
+                state: PieceState.King
+            },
+        };
+    }
+
+    setBoard(newBoard);
+    setCurrentPiece(null);
+    setUpdateBoard(!updateBoard);
+    switchSide(activeSide, setActiveSide);
+};
+
+const beatPiece = (
+    board: ICell[],
+    currentPiece: ICell,
+    selectedPiece: ICell,
+    updateBoard: boolean,
+    activeSide: PieceColor,
+    beatByWhite: ICell[],
+    beatByBlack: ICell[],
+    setCurrentPiece: (piece: ICell | null) => void,
+    setActiveSide: (activeSide: PieceColor) => void,
+    setBoard: (board: ICell[]) => void,
+    setUpdateBoard: (updateBoard: boolean) => void,
+    setCanBeatAgain: (canBeatAgain: boolean) => void,
+    canBeatPiece?: boolean,
+): void => {
+    const newBoard = Object.assign(board);
+
+    const currentPosition = newBoard.findIndex((item: ICell) => (
+        item.row === currentPiece?.row && item.col === currentPiece?.col)
+    );
+
+    const newPosition = newBoard.findIndex((item: ICell) => (
+        item.row === selectedPiece?.row && item.col === selectedPiece?.col)
+    );
+
+    let positionBetween;
+    let isWillBeKing = currentPiece?.piece?.state === PieceState.Man &&
+        isShouldBecomeKing(currentPiece, selectedPiece);
+
+    if (currentPiece?.piece?.state === PieceState.Man || canBeatPiece) {
```

Comment:
why do you have `currentPiece?.piece?` here? Can `currentPiece` not be defined?

Reply:
Actually, not. But in my interface I have that it can be Piece or null. So TS is warning me that it possibly can be null